### DISCUSSION TITLE
Opengrok UI doesn't show file contents for changesets for branched files

### DISCUSSION THREAD

**Discussion by jkjha** [oracle/opengrok#4369] [Q&A]
Its regarding the Perforce repository, where it seems to be an issue occurring with most of the files. This problem could be related to a configuration error or something that I think can be resolved. I am currently using a Docker version of OpenGrok, where the source code is mounted on a container.

Let's consider the file "example.h" as an example. When I open this file in OpenGrok and click on "history," I can view all the available historical data for the file. I can even select specific changesets, which takes me to the file at that particular changeset. However, for some older changesets (more than two years old), the page appears blank.

Interestingly, there are no error messages in the log, and it returns an exit code of 0.

```
<record>
  <date>2023-06-06T11:49:08.896488Z</date>
  <millis>1686052148896</millis>
  <nanos>488000</nanos>
  <sequence>259548</sequence>
  <logger>org.opengrok.indexer.util</logger>
  <level>FINE</level>
  <class>org.opengrok.indexer.util.Executor</class>
  <method>exec</method>
  <thread>2462</thread>
  <message>Finished command [/usr/bin/p4 print -q $'example.h@8018060'] in directory '/opengrok/src/folder_name/' with exit code 0</message>
</record>
```

Any idea what could be going wrong or any debugging tips?

**Comment by jkjha** [comment]
More debugging that I did:

When I run following command from the container shell for new changeset, it prints the file correctly:

```
p4 print -q 'example.h@10668717'  // new changeset

//file content
```

But when I run same command for old changeset (around 2 years old), I get error:
 
```
p4 print -q 'example.h@8173904'  // old changeset
example.h@8173904 - no file(s) at that changelist number.
```

But this file exists at that changeset.

**Reply by jkjha** [reply]
This seems related to branched file.
The file has been branched from "stage" branch to "main" branch.
While it is good that opengrok shows the history of the file before the branch, but history is useless because clicking on any revision before the branch returns an empty file.

Any suggestion, if we can handle this?

**Reply by vladak** [reply]
Not from my end as I've never worked with Perforce before and so this is and will likely be outside my capacity.

**Comment by jkjha** [comment]
Documenting all the details here... to open a feature request or request for contribution from p4 experts.

When a user clicks on History link of a file, it runs following command on backend and based on output it displays the result:

`p4 filelog -slti example.h` (from the project directory)

Above commands returns historical data which UI formats to display on UI. However, UI ignores some field from above output such as:

```
<changeset info>  revision 1
... ... moved from //depot/level1/main/lib/example.h#1,#145
//depot/level1/main/lib/example.h
<changeset info>  revision 145

...
...
some more changeset info

... ... moved from //depot/level1/stage/lib/example.h#1,#145
//depot/level1/stage/lib/example.h

```

These are the point where the file was moved from one branch to other branch. So instead of ignoring this, if we can keep this value in some variable and while getting the content of file with old branch, if we can use it then it can display the result.

So, instead of running, 

[/usr/bin/p4 print -q $'example.h@8018060'] in directory '/opengrok/src/project_name/'

run 

[/usr/bin/p4 print -q '<full_depot_path>' ] based on the above info, then it can give the file content. (full depot path available from above history info)

Changing branch is very common in perforce and so this can solve a big issue for user. Although I'm not familiar with Java, I will attempt to explore the code to see if I can implement these modifications on my local machine. In the meantime, I welcome any Perforce experts to provide their insights.
